ui/src/time_picker.tsx

```diff
@@ -0,0 +1,269 @@
+  placeholder?: S
+  /** The time value in hh:mm format. E.g. '10:30', '14:25', '23:59', '00:00' */
+  value?: S
+  /** True if this field is disabled. */
+  disabled?: B
+  /** The width of the time picker, e.g. '100px'. Defaults to '100%'. */
+  width?: S
+  /** True if the component should be visible. Defaults to True. */
+  visible?: B
+  /** True if the form should be submitted when the time is selected. */
+  trigger?: B
+  /** True if this is a required field. Defaults to False. */
+  required?: B
+  /** True if time picker should use a 12-hour time format. Defaults to True. */
+  time_format_12h?: B
+  /** The minimum allowed time value in hh:mm format. E.g.: '08:00', '13:30' */
+  min?: S
+  /** The maximum allowed time value in hh:mm format. E.g.: '15:30', '00:00' */
+  max?: S
+  /** Limits the available minutes to select from. One of `1`, `5`, `10`, `15`, `20`, `30` or `60`. Defaults to `1`. */
+  minutes_step?: U
+}
+
+const
+  css = stylesheet({
+    toolbar: {
+      paddingTop: 16,
+      paddingLeft: 16
+    },
+    toolbarTime: {
+      fontSize: 26,
+      cursor: 'pointer',
+      $nest: {
+        '&:hover': {
+          backgroundColor: cssVar('$neutralLighter'),
+          color: cssVar('$neutralDark')
+        }
+      }
+    },
+    toolbarText: { fontSize: 26 },
+    toolbarLabel: { maxWidth: '70%' }
+  }),
+  popoverProps: Partial<PopperProps> | undefined = {
+    placement: 'bottom-start',
+    sx: {
+      '& .MuiPaper-root': {
+        borderRadius: '2px',
+        boxShadow: `${cssVar('$text1')} 0px 6.4px 14.4px 0px, ${cssVar('$text2')} 0px 1.2px 3.6px 0px`
+      },
+      '& .MuiTypography-root': {
+        fontFamily: 'Inter',
+      }
+    }
+  }
+
+// Type 'any' instead of 'Date' because of warning when TimePicker is lazy loaded.
+type ToolbarProps = { params: BaseToolbarProps<any, any | null>, label: S | undefined, switchAmPm: () => void }
+type TextInputProps = {
+  props: {
+    onClick?: React.MouseEventHandler<HTMLInputElement | HTMLTextAreaElement>,
+    onChange?: React.FormEventHandler<HTMLInputElement | HTMLTextAreaElement>,
+    value?: Date | null,
+    placeholder?: S,
+    label?: S,
+    required?: B,
+    time_format_12h?: B,
+    errorMsg?: S,
+    error?: B,
+    disabled?: B
+  }
+}
+
+const
+  ThemeProvider = React.lazy(() => import('@mui/material/styles').then(({ ThemeProvider }) => ({ default: ThemeProvider }))),
+  LocalizationProvider = React.lazy(() => import('@mui/x-date-pickers').then(({ LocalizationProvider }) => ({ default: LocalizationProvider }))),
+  TimePicker = React.lazy(() => import('@mui/x-date-pickers').then(({ TimePicker }) => ({ default: TimePicker }))),
+  parseTimeStringToDate = (time: S) => new Date(`2000-01-01T${time.slice(0, 5)}:00`),
+  formatDateToTimeString = (date: Date, time_format_12h: B = true) => date.toLocaleString('en-US', { hour: '2-digit', minute: '2-digit', hourCycle: time_format_12h ? 'h12' : 'h23' }),
```

Comment:
Times like 24:00, 24:56 instead of 00:00, 00:56 look definitely odd to me and I've read it is used just rarely. In time of writing this code I was even thinking it is a bug but then I found out that this is called 24 hour cycle. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/hourCycle 
So I've used 23 hour life cycle to show the time in the most common way.

Even in the library we are using for time picker we can't select number 24 since there is 00 instead. Do we really want to support 24:xx times? I don't think it will be a common use case.

But you are right there is an inconsistency. According to current implementation of the time picker API should accept `12` and `23`. But now there's a question - which one to support/use? `24` or `23`?

Reply:
> Times like 24:00, 24:56 instead of 00:00, 00:56 look definitely odd to me

It seems like there is a difference between 24h **cycle** and 24h **time format**.

If we want to expose the cycle, not the format, IMO we should follow the implementation meaning both exposing and using the same hour cycle (exposing `23` and using `23` and so on). Otherwise, I am not sure how we would handle future requests for the `24h` cycle if we already used it for the `23` if that makes sense.

wdyt @lo5 ^^ ?

Btw https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/hourCycle has quite [low support](https://caniuse.com/?search=Intl.Locale) so I would suggest looking at different way of getting the right output (maybe newly imported DateFns has something with better support).

Reply:
Use `hour_format='12'` and `hour_format='24'` - that should keep things simple.